src/input-states.js

```diff
@@ -0,0 +1,289 @@
+ * @param portName the InPort name on this component 
+ */
+function getPortState(component, vnid, portName) {
+    var port = component.inPorts[portName];
+    if (port.isMulti()) return getPortStateArray(component, vnid, portName);
+    else if (component.vnis && component.vnis[vnid] && component.vnis[vnid][portName])
+        return component.vnis[vnid][portName];
+    else if (vnid) return getPortState(component, '', portName);
+    else return undefined;
+}
+
+/**
+ * Returns an array of the port states, one item for each attached socket. When
+ * all the port states are not need this is a convient way to
+ * retrieve just the port state that is needed.
+ * @param component a component facade
+ * @param vnid an identifier that distinguishes the set of states
+ * @param portName the InPort name on this component 
+ */
+function getPortStateArray(component, vnid, portName) {
+    var port = component.inPorts[portName];
+    var state = component.vnis && component.vnis[vnid] && component.vnis[vnid][portName];
+    if (!isMultiPort(port)) throw Error("This port is not addressable/multi: " + portName);
+    else if (!_.isArray(state)) return [];
+    else return port.listAttached().map(function(socketIndex){
+        return getPortStateIndex(component, vnid, portName, socketIndex);
+    });
+}
+
+/**
+ * Returns the state of this port for an identified socket. This can be used to
+ * look up the last state when checking to see if the state has changed.
+ * @param component a component facade
+ * @param vnid an identifier that distinguishes the set of states
+ * @param portName the name of the addressable/multi port for this component
+ * @param sourceNode the upstream component nodeId
+ * @param sourcePort the upstream component port name
+ */
+function getPortStateEntry(component, vnid, portName, sourceNode, sourcePort) {
+    if (!isMultiPort(component.inPorts[portName]))
+        throw Error("This port is not addressable/multi: " + portName);
+    var socketIndex = findSocketIndex(component, portName, sourceNode, sourcePort);
+    return getPortStateIndex(component, vnid, portName, socketIndex);
+}
+
+/**
+ * Returns the state of this port for the given socketIndex. This can be used to
+ * look up the last state when checking to see if the state has changed.
+ * Note that the socketIndex is not neccessarily the same as the array index
+ * @param component a component facade
+ * @param vnid an identifier that distinguishes the set of states
+ * @param portName the name of the addressable/multi port for this component
+ * @param socketIndex the socketIndex (or socketId) provided to a port data event
+ */
+function getPortStateIndex(component, vnid, portName, socketIndex) {
+    if (!isMultiPort(component.inPorts[portName]))
+        throw Error("This port is not addressable/multi: " + portName);
+    component.vnis = component.vnis || {};
+    component.vnis[vnid] = component.vnis[vnid] || {};
```

Comment:
I think it would be better to use a generic ensureVniExists function here, since this is likely to be a common need:
<pre>
var vni = ensureVniExists(component, vnid);
</pre>